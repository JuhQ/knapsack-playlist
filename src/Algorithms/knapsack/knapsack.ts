import Queue from "../../Datastructures/queue/queue";

/**
 * The recursive algorithm will call this algorithm to try to get the videos with the best ratings
 * The goal is to get a playlist which has rating as close to 5/5 as possible
 * @param list array of object
 * @param maxLength number
 * @param index number
 */
const recursiveKnapsackRating = (
  list: Queue,
  maxLength: number,
  index: number
): Queue => {
  if (index === -1) {
    return new Queue();
  }

  const currentVideo = list.at(index);
  const nextVideoWithSameLength = recursiveKnapsackRating(
    list,
    maxLength,
    index - 1
  );

  if (maxLength < currentVideo.seconds) {
    return nextVideoWithSameLength;
  }

  const nextVideoWithSmallerLength = recursiveKnapsackRating(
    list,
    maxLength - currentVideo.seconds,
    index - 1
  );

  nextVideoWithSmallerLength.enqueue(currentVideo);

  if (
    nextVideoWithSameLength.averageRating() >
    nextVideoWithSmallerLength.averageRating()
  ) {
    return nextVideoWithSameLength;
  }

  return nextVideoWithSmallerLength;
};

/**
 * The actual recursive implementation of the algorithm
 * @param list array of object
 * @param maxLength number
 * @param index number
 */
const recursiveKnapsack = (
  list: Queue,
  maxLength: number,
  index: number
): Queue => {
  if (index === -1) {
    return new Queue();
  }

  const currentVideo = list.at(index);
  const nextVideoWithSameLength = recursiveKnapsackRating(
    list,
    maxLength,
    index - 1
  );

  if (maxLength < currentVideo.seconds) {
    return nextVideoWithSameLength;
  }

  const nextVideoWithSmallerLength = recursiveKnapsackRating(
    list,
    maxLength - currentVideo.seconds,
    index - 1
  );

  nextVideoWithSmallerLength.enqueue(currentVideo);

  if (
    nextVideoWithSameLength.seconds() > nextVideoWithSmallerLength.seconds()
  ) {
    return nextVideoWithSameLength;
  }

  return nextVideoWithSmallerLength;
};

/**
 * Knapsack algorithm.
 * This is an recursive implementation of the algorithm, which will call itself until maximum size of the list is reached
 * @param list array of objects with following data structure: {id: string, title: string, seconds: number}
 * @param maxLength number
 */
const knapsack = (list: Queue, maxLength: number): Queue => {
  const queue = new Queue();
  const listLength = list.length();

  // if no values given, return empty list immediately
  if (maxLength <= 0 || listLength === 0) {
    return queue;
  }

  // if the given list already fits into the desired playlist length, just return the list
  if (list.seconds() <= maxLength) {
    // eslint-disable-next-line no-plusplus
    for (let i = 0; i < listLength; i++) {
      queue.enqueue(list.at(i));
    }

    return queue;
  }

  return recursiveKnapsack(list, maxLength, listLength - 1);
};

export default knapsack;
