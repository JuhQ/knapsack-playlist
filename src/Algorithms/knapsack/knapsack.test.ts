import Queue from "../../Datastructures/queue/queue";
import YoutubeMusic from "../../Models/Youtube/Youtube";
import generateTestData from "../../Utils/testHelpers";
import knapsack from "./knapsack";

describe("knapsack algorithm", () => {
  it("should return empty list when given weight is zero", () => {
    const result = knapsack(new Queue(), 0);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return empty list when given list is empty", () => {
    const result = knapsack(new Queue(), 1000);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return empty list when given list is empty and weight is zero", () => {
    const result = knapsack(new Queue(), 0);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return empty list when given list is empty and weight is negative", () => {
    const result = knapsack(new Queue(), -10);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return the only item if only one item in the list and the item fits", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 123, id: "test" });
    const result = knapsack(list, 123);
    expect(result.all().getAsArray()).toEqual([
      { title: "test", seconds: 123, id: "test" },
    ]);
  });

  it("should return empty list when the list contains data but the weight limit is zero", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 123, id: "test" });
    const result = knapsack(list, 0);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return empty list when the list contains data but the weight limit is negative", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 123, id: "test" });
    const result = knapsack(list, -1);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return empty list when the list contains large amount of data but the weight limit is negative", () => {
    const result = knapsack(generateTestData(10000), -1);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return empty list if all the videos are shorter than given length", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 123, id: "test" });
    const result = knapsack(list, 1);
    expect(result.all().getAsArray()).toEqual([]);
  });

  it("should return a list which only contains the first element which fits, when all the elements have same size and the weight limit matches the size", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 123, id: "test" });
    list.enqueue({ title: "test2", seconds: 123, id: "test2" });
    const result = knapsack(list, 123);
    expect(result.all().getAsArray()).toEqual([
      { title: "test", seconds: 123, id: "test" },
    ]);
  });

  it("should return a list which only contains elements which combined, fit into the given weight", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 3, id: "test" });
    list.enqueue({ title: "test2", seconds: 2, id: "test2" });
    const result = knapsack(list, 5);
    expect(result.all().getAsArray()).toEqual([
      { title: "test", seconds: 3, id: "test" },
      { title: "test2", seconds: 2, id: "test2" },
    ]);
  });

  it("should return a list which has length of 5, leaving one of the items out of the result", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 3, id: "test" });
    list.enqueue({ title: "test2", seconds: 2, id: "test2" });
    list.enqueue({ title: "test3", seconds: 1, id: "test3" });
    const result = knapsack(list, 5);
    expect(result.all().getAsArray()).toEqual([
      { title: "test", seconds: 3, id: "test" },
      { title: "test2", seconds: 2, id: "test2" },
    ]);
  });

  it("should return a list which contains all the items, because the weight is larger than the sum of item weights", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 3, id: "test" });
    list.enqueue({ title: "test2", seconds: 2, id: "test2" });
    const result = knapsack(list, 6);
    expect(result.all().getAsArray()).toEqual([
      { title: "test", seconds: 3, id: "test" },
      { title: "test2", seconds: 2, id: "test2" },
    ]);
  });

  it("should return a list which contains all the items, because the weight is 1000 times larger than the sum of item weights", () => {
    const list = new Queue();
    list.enqueue({ title: "test", seconds: 3, id: "test" });
    list.enqueue({ title: "test2", seconds: 2, id: "test2" });
    const result = knapsack(list, 5000);
    expect(result.all().getAsArray()).toEqual([
      { title: "test", seconds: 3, id: "test" },
      { title: "test2", seconds: 2, id: "test2" },
    ]);
  });

  describe("short playlists with full music catalog", () => {
    const catalog = YoutubeMusic();

    it("should return a one minute playlist", () => {
      const result = knapsack(catalog, 60);
      expect(result.seconds()).toBe(45);
    });
    it("should return a two minute playlist", () => {
      const result = knapsack(catalog, 120);
      expect(result.seconds()).toBe(88);
    });
    it("should return a three minute playlist", () => {
      const result = knapsack(catalog, 180);
      expect(result.seconds()).toBe(176);
    });
    it("should return a four minute playlist", () => {
      const result = knapsack(catalog, 240);
      expect(result.seconds()).toBe(223);
    });
    it("should return a five minute playlist", () => {
      const result = knapsack(catalog, 300);
      expect(result.seconds()).toBe(284);
    });
  });

  it("should return one hour long playlist", () => {
    // take 20 videos from the list and try to generate one hour of music

    const list = new Queue();
    list.merge(YoutubeMusic().slice(-20));

    const playlist = knapsack(list, 3600);

    expect(playlist.seconds()).toBe(3431);
    expect(playlist.all().getAsArray()).toMatchSnapshot();
  });

  // these test cases are here to show the difference in playlist length when selecting different subset of videos
  // as we can see from the expected results, they do not match 100% with the desired playlist length. This will probably change once the playlist contains more items
  describe("variations on playlist", () => {
    describe("one hour playlist", () => {
      it("should return one hour long playlist, on a slice fom 0 to 20 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(0, 20));
        const playlist = knapsack(list, 3600);
        expect(playlist.seconds()).toBe(3531);
      });
      it("should return one hour long playlist, on a slice fom 20 to 40 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(20, 40));
        const playlist = knapsack(list, 3600);
        expect(playlist.seconds()).toBe(3588);
      });
      it("should return one hour long playlist, on a slice fom 40 to 60 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(40, 60));
        const playlist = knapsack(list, 3600);
        expect(playlist.seconds()).toBe(3566);
      });
      it("should return one hour long playlist, on a slice fom 60 to 80 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(60, 80));
        const playlist = knapsack(list, 3600);
        expect(playlist.seconds()).toBe(3437);
      });
    });

    describe("two hour playlist", () => {
      it("should return two hour long playlist, on a slice fom 0 to 20 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(0, 20));
        const playlist = knapsack(list, 3600 * 2);
        expect(playlist.seconds()).toBe(5198);
      });
      it("should return two hour long playlist, on a slice fom 20 to 40 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(20, 40));
        const playlist = knapsack(list, 3600 * 2);
        expect(playlist.seconds()).toBe(5852);
      });
      it("should return two hour long playlist, on a slice fom 40 to 60 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(40, 60));
        const playlist = knapsack(list, 3600 * 2);
        expect(playlist.seconds()).toBe(5948);
      });
      it("should return two hour long playlist, on a slice fom 60 to 80 from music list", () => {
        const list = new Queue();
        list.merge(YoutubeMusic().slice(60, 80));
        const playlist = knapsack(list, 3600 * 2);
        expect(playlist.seconds()).toBe(6994);
      });
    });
  });

  describe("exact matches", () => {
    const matches = [
      { seconds: 10, weight: 100 },
      { seconds: 20, weight: 200 },
      { seconds: 30, weight: 300 },
      { seconds: 40, weight: 400 },
      { seconds: 50, weight: 500 },
    ];

    matches.forEach(({ seconds, weight }) => {
      it(`should return a playlist which is ${weight} seconds long, given small enough items`, () => {
        const items: Queue = generateTestData(10, seconds);
        const playlist = knapsack(items, weight);
        expect(playlist.seconds()).toBe(weight);
      });
    });
  });

  // test commented because it's a bit slow
  // it("should return a two hour long playlist", () => {
  //   // take 30 videos from the list and try to generate two hours of music
  //   const list = new Queue();

  //   list.merge(YoutubeMusic().slice(-30));
  //   const playlist = knapsack(list, 3600 * 2);

  //   expect(playlist.seconds()).toBe(7167);
  // });
});
