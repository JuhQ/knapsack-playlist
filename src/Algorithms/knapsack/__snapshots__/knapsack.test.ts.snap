// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`knapsack algorithm should return one hour long playlist 1`] = `
Array [
  Object {
    "id": "VJboSby7nW0",
    "seconds": 297,
    "title": "Russian Rammstein - Du Riechst So Gut",
  },
  Object {
    "id": "10JDA8SvwX8",
    "seconds": 235,
    "title": "Rammstein - Stripped (Official Video)",
  },
  Object {
    "id": "xkEju-deJsE",
    "seconds": 202,
    "title": "LINDEMANN - Ach so gern (One Shot Video)",
  },
  Object {
    "id": "EOnSh3QlpbQ",
    "seconds": 246,
    "title": "Rammstein - Ich Will (Official Video)",
  },
  Object {
    "id": "CSvFpBOe8eY",
    "seconds": 207,
    "title": "System Of A Down - Chop Suey! (Official Video)",
  },
  Object {
    "id": "U7IUA8W8xuM",
    "seconds": 331,
    "title": "RAGE AGAINST THE MACHINE - KILLING IN THE NAME - HARDCORE WORLDWIDE (OFFICIAL VERSION HCWW)",
  },
  Object {
    "id": "gNdnVVHfseA",
    "seconds": 228,
    "title": "Rammstein - Mutter (Official Video)",
  },
  Object {
    "id": "uT3SBzmDxGk",
    "seconds": 299,
    "title": "2CELLOS - Thunderstruck [OFFICIAL VIDEO]",
  },
  Object {
    "id": "B9FzVhw8_bY",
    "seconds": 250,
    "title": "The Dead South - In Hell I'll Be In Good Company [Official Music Video]",
  },
  Object {
    "id": "jM8dCGIm6yc",
    "seconds": 397,
    "title": "The HU - Wolf Totem (Official Music Video)",
  },
  Object {
    "id": "cegdR0GiJl4",
    "seconds": 313,
    "title": "Die Antwoord - Enter The Ninja (Explicit Version) (Official Video)",
  },
  Object {
    "id": "v2AC41dglnM",
    "seconds": 293,
    "title": "AC/DC - Thunderstruck (Official Video)",
  },
  Object {
    "id": "IxuEtL7gxoM",
    "seconds": 238,
    "title": "Rammstein - Ich Tu Dir Weh (Official Video)",
  },
  Object {
    "id": "cTCXDdHUFE8",
    "seconds": 18,
    "title": "The Passion for Music â€“ Short",
  },
]
`;
